### PR TITLE
Client Support for Docker Registry HTTP API V2

### PR DESCRIPTION
**Pull request by stevvooe**
This PR updates Docker core to support pushes and pulls of images from registries that support Docker Registry HTTP API V2, as proposed in docker/docker#9015. Specifically, functionality has been added to support the V2 API while minimizing changes to the end user experience. The reasons behind this move are to improve security and access control by moving to content addressable layers and using sign-able manifests.

This PR affects the following behavior in docker core:
- `docker login`
- Pull of images.
- Push of images

While the distribution side of docker core may need some refactoring work, the goal of this particular PR is to provide equivalent functionality with the new registry API. Use of a V2 registry should be transparent to the user.

This set of changes will be supported by some accompanying PR:
- Docker Registry HTTP V2 specification (unsubmitted, based on docker/docker#9015).
- docker logout (unsubmitted)
## Overview

Images may now be pushed and pulled from registries, public and private, that implement the Docker Registry HTTP API V2.  Using `registry/v2.URLBuilder`, the client code has been ported to use the URI layout described in the proposal. Definitions for routes and error codes from [next generation docker registry](https://github.com/docker/docker-registry/tree/next-generation) are included. The rest of the changes are devoted to authentication handling and modifying the flow to support image push. 

During login, if a registry supports the V2 api, the client will use the base endpoint (“/v2/”) to solicit an authentication challenge. The challenge is then used to login to the appropriate authentication provider. The results are tested against the base endpoint to confirm the credentials are correct. After login, authorization of operations on the registry is maintained using challenges and the stored credentials. If a challenge is issued during an operation against the registry, the client attempts to re-authorize the operation, if possible.

Use of the new image format, covered in docker/docker#8093, is effectively transparent. Image manifests are created and signed transparently. This paves the way for future image verification features. A new API endpoint has been added to the docker engine to get the image manifest for an image. This endpoint can be used to get the image manifest from a daemon, sign it on the client, and push the signed image. The push API endpoint has also been updated to allow a signed manifest to be sent in the post body.

Pulling images is roughly the same as before, with most modifications to this code path consisting of URL updates and authentication management. For pushing images, the POST + PUT monolithic flow is implemented as proposed.
## Testing

There are numerous cases under which the new registry may be used. We suggest that reviewers do take the time to test cases that they care about directly. To ensure that this PR is somewhat ready, we have tested the following combinations against the [next generation registry](https://github.com/docker/docker-registry/tree/next-generation):
- “Production” registry with token auth
- Private registry with token auth
- Private registry with basic auth
- Private registry with no auth

If you need help with reproducing these scenarios, please contact us directly.

### CODE REVIEW

File: registry/session_v2.go [hunk 1]
```diff
@@ -267,80 +192,82 @@ func (r *Session) GetV2ImageBlobReader(imageName, sumType, sum string, token []s
...
-	req, err := r.reqFactory.NewRequest(method, routeURL.String(), blobRdr)
+	log.Debugf("[registry] Calling %q %s", method, location)
+	req, err = r.reqFactory.NewRequest(method, location, blobRdr)
 	if err != nil {
-		return "", err
+		return err
 	}
-	setTokenAuth(req, token)
-	res, _, err := r.doRequest(req)
+	queryParams := url.Values{}
```
Comment (stevvooe) [inline registry/session_v2.go]: These parameter additions need to be additive. With this approach, any query parameters in the `Location` header will be destroyed. The code should be something like this:

```
query := req.URL.Query()
query.Add("digest", sumType+":"+sumStr)
req.URL.RawQuery = query.Encode()
```

**Comment by LK4D4** [comment]
```
docker pull -a registry-staging.docker.io/lk4d4/ipv6
Pulling repository registry-staging.docker.io/lk4d4/ipv6
FATA[0001] HTTP code: 405
```
pulling without `-a`
Also 405 pretty weird code to see :)

**Comment by LK4D4** [comment]
also seems 405 everywhere, on nonexistent images too.

**Comment by stevvooe** [comment]
@LK4D4 What are the corresponding daemon logs? A 405 is returned when no handlers are matched by a particular method (which is an implementation detail of gorilla/mux).

**Comment by jfrazelle** [comment]
theres an empty log somewhere:
```
DEBU[0313] () does not match latest (id: 7ee8afe2672221108b47b95a284eefb22cf37cd69689f7720c9b78baa9d3796e), skipping 
DEBU[0313] () does not match latest (id: d86c3be7a1464c543442f8f8ae927317cf264d34523c33c33320620a8f3e152c), skipping 
DEBU[0313] () does not match latest (id: 55134ad0e5aa8251485fa051029320fb9671db410277f238c023bb810908802a), skipping 
DEBU[0313] () does not match latest (id: 2a29728ad7af2794b82bc19fcfa40ef0ec3cafb0f7c68b10006ef259acd8745a), skipping 
DEBU[0313] () does not match latest (id: 54bbc38e71e0db5e6b234058b7fcdc737fb3cea6549e9336bf28fc0bf705d14e), skipping 
DEBU[0313] () does not match latest (id: f57aea76eca6196f72c25e59b28d251c81c9493243dceea5c0636714a4443a78), skipping 
DEBU[0313] () does not match latest (id: aa9a7c60c1cccbbdc02d2cd93fb29b72d7cae68cb9669f817e1eff77436e81b3), skipping 
DEBU[0313] () does not match latest (id: 3687839bafb21be92a5b15d611c8b21d0a6fb82b1e88694bbc4057b723316384), skipping 
DEBU[0313] () does not match latest (id: cf8dc907452c970224551599da573c9e32897fc65286d942625c4c86dabd680d), skipping 
DEBU[0313] () does not match latest (id: 2b6070780f1aa4db80ff829cf6ac2c882fe9f886a6f4a8a3eb8abb3e04286977), skipping 
DEBU[0313] () does not match latest (id: 03ab93e4fba877ed4aa565d0e82963b286c422140f933a89948db3701ee38c36), skipping 
DEBU[0313] () does not match latest (id: 3a6f1deb5d2ac3dc23140769fc5510dee69110bfbea6f0812a0c75148ec0cd42), skipping 
DEBU[0313] () does not match latest (id: 39bf9508303d5d793f6cae5b2021ab68bd0954a9f588f3d734361e82204b88dc), skipping 
DEBU[0313] () does not match latest (id: 79596999b8d3f803af63e38021fc30f00218ccebb3ee1ecc802c0592caf07578), skipping 
DEBU[0313] () does not match latest (id: 13fed73e576666252c5ae0a947168404667b639c289bf3505be4cfbf899de65f), skipping 
DEBU[0313] () does not match latest (id: 80a54c830e1012cc25a31af21c56f2f165017699810cfcfcf45b4c51bc17b9f1), skipping 
DEBU[0313] () does not match latest (id: a1c03985d5a4133d463b7ed262c039c46f0daa94f3e71d6ad815a15bef0aa0f2), skipping 
DEBU[0313] () does not match latest (id: 2b750b856e28e48098dfda59427f25a6f76265479954e67467b79b84088a3122), skipping 
DEBU[0313] () does not match latest (id: 28774ca5b76efccd56f9b74b4e3f7be1ea62dc4fb48a30ce3786d946e089594d), skipping 
DEBU[0313] () does not match latest (id: e314931015bd259897b87925c7756c33235ae3a7a28eb4accf83169a5b862a5d), skipping 
DEBU[0313] () does not match latest (id: 4b30a90166db902860ee80f24681b8e5d6de11fc6ecf8a786b65929d53686d0f), skipping 
DEBU[0313] () does not match latest (id: 589faf326e3124d8ddc66c5bdc91d923fa2184b2440b2836aefd095348b5094e), skipping 
DEBU[0313] () does not match latest (id: e267bd800f9feda4564db7583bf0aeb7032c3af6f70008d9329f01d5753946f3), skipping 
DEBU[0313] () does not match latest (id: 60bebdd2ecf9683851859d705d1df77db352e502300d9742825dd7dca483a47b), skipping 
DEBU[0313] () does not match latest (id: 511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158), skipping 
DEBU[0313] () does not match latest (id: 78096d7feb0cfc711454084d102bc47dba4537ee94065cdadce44d30117ce228), skipping 
DEBU[0313] () does not match latest (id: c5881f11ded97fd2252adf93268114329e985624c5d7bb86e439a36109d1124e), skipping 
DEBU[0313] () does not match latest (id: 8955ec51479c1b4f8014a99321e4eef3283f2c0e77b5d20708eb8e74d49bf126), skipping 
DEBU[0313] () does not match latest (id: 4daf60352bebcd87df6a0eb5a66ed1921b84dde0150d529cd492eb37946b6e01), skipping 
DEBU[0313] () does not match latest (id: 4ecde1041742d1199fa54b99fcbb27f5fb49d142725ac7d9ec04ba3b5e725bbb), skipping 
DEBU[0313] () does not match latest (id: a37d6f68bd72978d5bc78d18af1b4e1380b3b811697713633fb2ff9d03a101ef), skipping 
DEBU[0313] () does not match latest (id: 707b6b3832943170244d6882511d3bfdf6f767b4c3a7b7d4dcb5c33d8fdb6f76), skipping 
DEBU[0313] () does not match latest (id: 4b908a3535f1b1a56183f1696917b02ae14f18bd76794c66d81ea1da84c57402), skipping 
DEBU[0313] () does not match latest (id: bf666b035f5c27ea69995a0e749213f2457b96480776bc059e35ebe5b9a5d44f), skipping 
DEBU[0313] () does not match latest (id: 88b37d759e1f8b1714d62e1bd5d8504e09f177750d08f18040294859104cefbd), skipping 
DEBU[0313] () does not match latest (id: fff99b7334efdac07be91f5bceb86606797e0e1393814599d3fdd39b83403c7c), skipping 
DEBU[0313] () does not match latest (id: 26bf7cebb55e49641b97910ceaa1fe766a2749ea1f7bf83c71e96c606c74abb5), skipping 
DEBU[0313] () does not match latest (id: f6168aa9c50946507acd4c874d85d454ffe573e9d054bc062ed02cb357426ad8), skipping 
......
```
what should be in the `()`

**Comment by jfrazelle** [comment]
who is this log helping w the byte output?
```
DEBU[0385] [tar autodetect] n: [31 139 8 0 0 9 110 136 0 255]        
DEBU[0385] [tar autodetect] n: [118 97 114 47 108 105 98 47 100 112] 
```

**Comment by jfrazelle** [comment]
```
$ docker push jess/mutt
511136ea3c5a: Image push failed 
FATA[0012] Put ?digest=tarsum.v1%2Bsha256%3Ae3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855: unsupported protocol scheme "" 
```

**Comment by dmcgowan** [comment]
Fixed `pull -a` in last commit.

@jfrazelle what modifications were made to support `docker push jess/mutt`, does it work when explicitly specifying a v2 registry?

We need to figure out what the proper error handling will be when attempting to fallback to v1 on a v2 only private registry.  This is what is causing the 405.

**Comment by jfrazelle** [comment]
how do i explicitly specify as v2?

**Comment by LK4D4** [comment]
Also drone failed

**Comment by LK4D4** [comment]
Okay, because this will be used in production since `1.5` we definitely need tests. So at monday I and @icecrime will try to up local registry v2 in our image and try to write couple of tests and then you'll try to rebase and write as much tests as you can.

**Comment by BrianBland** [comment]
I found an issue with the progress reporting for push with this change which can be quite misleading.

It appears that the layers are pushed in parallel (which is awesome), but if a large layer (call it A) is followed by a small layer (B) in sequential order, there is a very large chance that layer B will print an "Image already exists" message.

Looking at the logs from the registry side, it is clear that B is actually pushed, but from the engine/CLI side of things, at the time that layer A is finished being pushed and the client progress printer gets to B, the engine actually checks the registry for the remote existence of B (which does now exists). Now, instead of displaying an "Image successfully pushed" message, we get the aforementioned misleading "Image already exists".

Certainly this would be fixed by displaying each of the progress bars at the same time, but in the interim we should at least changed this message.

**Comment by jfrazelle** [comment]
I'm going to swap my binary out with this and see what I run into in the next day

File: graph/manifest.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+		return nil, err
+	}
+	if localRepo == nil {
+		return nil, fmt.Errorf("Repo does not exist: %s", localName)
+	}
+
+	// Get the top-most layer id which the tag points to
+	layerId, exists := localRepo[tag]
+	if !exists {
+		return nil, fmt.Errorf("Tag does not exist for %s: %s", localName, tag)
```
Comment (icecrime) [inline graph/manifest.go]: Nitpick: I just noticed an extra space between name and tag

File: registry/auth.go [hunk 3]
```diff
@@ -36,6 +37,73 @@ type ConfigFile struct {
...
+	return &RequestAuthorization{
+		authConfig:       authConfig,
+		registryEndpoint: registryEndpoint,
+		resource:         resource,
+		scope:            scope,
+		actions:          actions,
+	}
+}
+
+func (auth *RequestAuthorization) getToken() (string, error) {
```
Comment (icecrime) [inline registry/auth.go]: This function has many `// TODO`, and is related to auth: what are the implications?

Comment (dmcgowan) [inline registry/auth.go]: The TODOs involve optimizing when a token is retrieved.  Because the tokens are short lived, they should not be cached indefinitely.  The optimization is cache for a short period of time.  The implication of not caching is a new token is retrieved for every registry request, which will be two per layer + a call for the manifest.

**Comment by dmcgowan** [comment]
Added some refactoring for the push and pull code which should make the diff easier to review

Also the issue mentioned by @BrianBland is not caused by parallel push (push is not parallel, only pull), rather pushing multiple empty layers.  When an empty layer is attempting to be pushed, it will see the layer as already existing since all empty layers have the same tarsum. This will appear to the user as layers already having been pushed even though the image has never been pushed.

**Comment by BrianBland** [comment]
Yeah sorry, I got a bit confused there.

**Comment by dmcgowan** [comment]
@LK4D4 @icecrime how can we get drone to use the correct image using the updated dockerfile?

@BrianBland it is confusing, meaning others might be confused as well.  I think the only possible way to make this particular case less confusing is by pointing out the layer is empty or skip the messaging for it entirely.

**Comment by icecrime** [comment]
Ping @jfrazelle, our Drone feeder.

**Comment by jfrazelle** [comment]
no we cant because drone/drone#1 and it would effect all other PRs, we can test manually. sorry.

File: Dockerfile [hunk 4]
```diff
@@ -130,6 +130,17 @@ RUN set -x \
 	&& git clone -b v1.2 https://github.com/russross/blackfriday.git /go/src/github.com/russross/blackfriday \
 	&& go install -v github.com/cpuguy83/go-md2man
 
+# Install registry
+COPY pkg/tarsum /go/src/github.com/docker/docker/pkg/tarsum
```
Comment (jfrazelle) [inline Dockerfile]: why?

Comment (jfrazelle) [inline Dockerfile]: shouldnt this just work? without this line?

File: Dockerfile [hunk 5]
```diff
@@ -130,6 +130,17 @@ RUN set -x \
...
+# REGISTRY_COMMIT gives us the repeatability guarantees we need
+# (so that we're all testing the same version of the registry)
+ENV REGISTRY_COMMIT 21a69f53b5c7986b831f33849d551cd59ec8cbd1
+RUN set -x \
+	&& git clone https://github.com/docker/distribution.git /go/src/github.com/docker/distribution \
+	&& (cd /go/src/github.com/docker/distribution && git checkout -q $REGISTRY_COMMIT) \
+	&& go get -d github.com/docker/distribution/cmd/registry \
+	&& go build -o /go/bin/registry-v2 github.com/docker/distribution/cmd/registry
```
Comment (jfrazelle) [inline Dockerfile]: this is only for tests right or are you using this in the code?

Comment (jfrazelle) [inline Dockerfile]: you could split adding the tests into a different PR from this which might be cool

File: Dockerfile (hunk 4 above)
Comment (tianon) [inline Dockerfile]: Try it, Jess. ;)

File: Dockerfile (hunk 5 above)
Comment (icecrime) [inline Dockerfile]: Yes, this is only for tests (we exec the `registry-v2` binary so we can `push`/`pull` to it).

TBH before this PR there was basically no tests (only 4 written, 3 of them being `t.Skip()`), so I'm not sure it's worth splitting.

File: Dockerfile (hunk 4 above)
Comment (jfrazelle) [inline Dockerfile]: but but but je suis le lazyyyy, jk jk i trust you

File: Dockerfile (hunk 5 above)
Comment (jfrazelle) [inline Dockerfile]: ok no bigs just wondering

File: Dockerfile (hunk 4 above)
Comment (tianon) [inline Dockerfile]: Then I'm not doing my job right.  It's always my fault, so you really
shouldn't trust me. ;)

**Comment by jfrazelle** [comment]
Things to note:
- Please test this PR by hand, it is only failing on drone because drone is not picking up the changes to the Dockerfile (see drone/drone#1)
- The entire test suite passes for me locally. :)
- If you would like to help test this PR and play with the binary below are links

**linux amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/amd64/docker-1.4.1-dev.sha256

**linux 386**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/386/docker-1.4.1-dev.sha256

**linux arm**
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/linux/arm/docker-1.4.1-dev.sha256

**darwin 386**
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/386/docker-1.4.1-dev.sha256

**darwin amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev
 http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/darwin/amd64/docker-1.4.1-dev.sha256

**freebsd 386**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/386/docker-1.4.1-dev.sha256

**freebsd amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/amd64/docker-1.4.1-dev.sha256

**freebsd arm**
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev.md5
http://jesss.s3.amazonaws.com/docker/pr9784/freebsd/arm/docker-1.4.1-dev.sha256

**windows 386**
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe.md5
http://jesss.s3.amazonaws.com/docker/pr9784/windows/386/docker-1.4.1-dev.exe.sha256

**windows amd64**
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe.md5
http://jesss.s3.amazonaws.com/docker/pr9784/windows/amd64/docker-1.4.1-dev.exe.sha256

**Comment by jfrazelle** [comment]
Also just because this is such a large change, I have hooked up @dmcgowan's branch to run on jenkins here https://jenkins.dockerproject.com/job/Docker%20PR%20Test/

**Comment by jfrazelle** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM
At least I tested locally :)

**Comment by icecrime** [comment]
LGTM! @jfrazelle can we let you proceed with the merge as we have to synchronize with Drone image update?

Thanks a lot to everyone involved in reviewing this <3

**Comment by stevvooe** [comment]
Thanks for getting this merged!
